mldatafind/find.py

```diff
@@ -23,86 +19,138 @@
-    out = []
-    for segment in segments:
-        duration = segment[1] - segment[0]
-        if duration <= chunk_size:
-            out.append(segment)
+        futures = OrderedDict()
+
+    def maybe_submit(current_memory, return_value=None):
+        # if we passed a future and we're not chunking,
+        # this means we're about to yield it, so subtract
+        # its memory footprint from our running total
+        if return_value is not None and chunk_size is None:
+            memory = futures.pop(return_value)
+            current_memory -= memory
+            return_value = return_value.result()
+        elif return_value is not None:
+            rm_futures.append(return_value)
+
+        # if our memory is currently full or we have no
+        # more segments to submit for loading, then
+        # short-circuit here
+        if current_memory > MEMORY_LIMIT or not segments:
+            return return_value
+
+        # check the next segment to see if we can load it
+        start, stop = segments.pop()
+        duration = stop - start
+
+        # if we're chunking our segments, return a
+        # generator of segments rather than
+        if chunk_size is not None:
+            if duration > chunk_size:
+                num_segments = int((duration - 1) // chunk_size) + 1
+                segs = []
+                for i in range(num_segments):
+                    end = min(start + (i + 1) * chunk_size, stop)
+                    seg = (start + i * chunk_size, end)
+                    segs.append(seg)
+            else:
+                segs = [seg]
+
+            # call this function recursively but with
+            # chunking turned off since we know that
+            # all the segments will have the right length
+            gen = data_generator(
+                exc,
+                segs,
+                loader,
+                channels,
+                chunk_size=None,
+                current_memory=current_memory,
+                retain_order=True,
+            )
+            futures.append(gen)
+            return return_value or gen
+
+        # if we're not chunking, submit this segment for loading
+        # TODO: should we check if this is going to put us over?
+        # The downside is that we'll never be over, and so the
+        # check at the top will have to be adjusted.
+        mem = utils._estimate_memory(len(channels), duration)
+        future = exc.submit(loader, channels, start, stop)
+        logging.debug(
+            "Submitted future to query {}s of data "
+            "and {:0.2f}GB of memory".format(duration, mem)
+        )
+
+        # record its memory footprint and future
+        current_memory += mem
+        futures[future] = mem
+        return return_value or future
+
+    while segments or futures:
+        # submit as many jobs as we can up front
+        while True:
+            if maybe_submit(current_memory, None) is None:
```

Comment:
Spent some time playing around with this. So I'm pretty sure this isn't actually updating `current_memory`, since floats are immutable. A hacky work around would be to make current_memory a one element list.

Reply:
Ahh duh of course, good catch. Yeah less than ideal but that's probably the most straightforward way to get this running. I can implement this today